src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -222,45 +236,57 @@ public UnfilteredRowIterator computeNext()
         /**
          * Returns the next available key contained by one of the keyRanges and selected by the queryController.
          * If the next key falls out of the current key range, it skips to the next key range, and so on.
-         * If no more keys acceptd by the controller are available, returns null.
+         * If no more keys acceptd by the controller are available, returns an empty list.
          */
-        private @Nullable PrimaryKey nextSelectedKeyInRange()
+        private List<PrimaryKey> nextSelectedKeyInRange()
         {
+            List<PrimaryKey> threadLocalNextKeys = nextKeys.get();
+            threadLocalNextKeys.clear();
             PrimaryKey key;
+
             do
             {
                 key = nextKeyInRange();
+
+                if (key == null)
+                    return Collections.emptyList();
             }
-            while (key != null && queryController.doesNotSelect(key));
-            return key;
+            while (queryController.doesNotSelect(key) || key.equals(lastKey));
```

Comment:
when entering the loop for the 1st time, `lastKey` is null. It will exist the loop. `threadLocalNextKeys` will only have one element. Is it expected?

Reply:
in my understanding, `nextSelectedKeyInRange()` always returns 1 single PrimaryKey

Reply:
So we don't _need_ `threadLocalNextKeys`, but I wanted to keep the return type of both suppliers (`nextSelectedKeyInRange()` and `nextSelectedKeysInPartition()`) the same to simplify `nextRowIterator()`, and using the thread-local avoids having to create a singleton collection for the one key.